activerecord/lib/active_record/core.rb

```diff
@@ -209,7 +209,7 @@ def find_by(*args) # :nodoc:
 
         keys = hash.keys
 
-        statement = cached_find_by_statement(keys) { |params|
+        statement = cached_find_by_statement("#{table_name}#{keys}") { |params|
```

Comment:
Nope. The `table_name=` method doesn't clear statement cache Unfortunately, but your comment gave me inspiration. I changed my approach.